derivation/shared/src/main/scala/io/circe/derivation/Configuration.scala

```diff
@@ -0,0 +1,96 @@
+package io.circe.derivation
+
+/**
+  * Configuration allowing customisation of the JSON produced when encoding, or
+  * expected when decoding. Can be passed to the [[JsonCodec]] annotation to
+  * allow customisation of derivation.
+  *
+  * The configuration also determines if either *both* encoder and decoder are
+  * derived - or if only one of them will be.
+  */
+sealed trait Configuration[A <: Configuration[A]] {
+
+  /** Transforms the names of type members in the JSON, allowing, for example,
+   *  formatting or case changes
+   */
+  def transformMemberNames: String => String
+
+  protected def getA(transformMemberNames: String => String): A
+
+  /** Creates a configuration which produces snake cased member names */
+  final def withSnakeCaseMemberNames: A =
+    getA(Configuration.snakeCaseMapping)
+
+  /** Creates a configuration which produces kebab cased member names */
+  final def withKebabCaseMemberNames: A =
+    getA(Configuration.kebabCaseMapping)
+}
+
+object Configuration {
+
+  /** Configuration allowing customisation of JSON produced when encoding or
+   *  decoding.
+   *
+   *  This configuration creates *both* encoder and decoder.
+   */
+  final case class Both(
+    transformMemberNames: String => String
+  ) extends Configuration[Both] {
+
+    protected final def getA(transformMemberNames: String => String) =
+      Both(transformMemberNames)
+  }
+
+  /** Configuration allowing customisation of JSON produced when encoding or
+   *  decoding.
+   *
+   *  This configuration **only** creates decoder.
+   */
+  final case class DecodeOnly(
+    transformMemberNames: String => String,
+  ) extends Configuration[DecodeOnly] {
+
+    protected final def getA(transformMemberNames: String => String) =
+      DecodeOnly(transformMemberNames)
+  }
+
+  /** Configuration allowing customisation of JSON produced when encoding or
+   *  decoding.
+   *
+   *  This configuration **only** creates encoder.
+   */
+  final case class EncodeOnly(
+    transformMemberNames: String => String
+  ) extends Configuration[EncodeOnly] {
+
+    protected final def getA(transformMemberNames: String => String) =
+      EncodeOnly(transformMemberNames)
+  }
+
+  /** Create a default configuration with both decoder and encoder */
+  val default: Both =
+    new Both(identity)
+
+  /** Create a default configuration with **only** encoder */
+  val encodeOnly: EncodeOnly =
+    new EncodeOnly(identity)
+
+  /** Create a default configuration with **only** decoder */
+  val decodeOnly: DecodeOnly =
+    new DecodeOnly(identity)
+
+  /** Snake case mapping for member names */
+  val snakeCaseMapping: String => String =
```

Comment:
there's already such code in the `renaming` package object

Reply:
Cool - didn't see that one -- I'll move the renames there :+1: